src/BizHawk.Emulation.Cores/Computers/Commodore64/Cartridge/Mapper0013.cs

```diff
@@ -11,126 +11,118 @@ namespace BizHawk.Emulation.Cores.Computers.Commodore64.Cartridge
+		private const byte DummyData = 0xFF;
 
-		private int _bankMask;
-		private int _bankNumber;
-
-		private int[] _currentBank;
+		private readonly byte[][] _banks = new byte[128][];
 
+		private readonly byte _bankMask;
+		private readonly int _bankCount;
+		private byte _bankNumber;
+		private byte[] _currentBank;
 		private bool _romEnable;
 
-		public Mapper0013(IList<int> newAddresses, IList<int> newBanks, IList<int[]> newData)
+		public Mapper0013(IEnumerable<CartridgeChip> chips)
 		{
-			var count = newAddresses.Count;
-
 			pinGame = true;
 			pinExRom = false;
 			_romEnable = true;
 
-			// build dummy bank
-			var dummyBank = new int[0x2000];
-			for (var i = 0; i < 0x2000; i++)
-			{
-				dummyBank[i] = 0xFF; // todo: determine if this is correct
-			}
+			// This bank will be chosen if uninitialized.
+			var dummyBank = new byte[BankSize];
+			dummyBank.AsSpan().Fill(DummyData);
+			_banks.AsSpan().Fill(dummyBank);
+			_bankMask = 0x00;
 
-			switch (count)
+			// Load in each bank.
+			var maxBank = 0;
+			foreach (var chip in chips)
 			{
-				case 16:
-					_bankMask = 0x0F;
-					_banks = new int[16][];
-					break;
-				case 8:
-					_bankMask = 0x07;
-					_banks = new int[8][];
-					break;
-				case 4:
-					_bankMask = 0x03;
-					_banks = new int[4][];
-					break;
-				default:
-					throw new Exception("This looks like a Domark/HES cartridge but cannot be loaded...");
+				// Maximum 128 banks.
+				if (chip.Bank is > 0x7F or < 0x00)
+					throw new Exception("Cartridge image has an invalid bank");
+				
+				// Addresses other than 0x8000 are not supported.
+				if (chip.Address != 0x8000)
+					continue;
+
+				// Bank wrap-around is based on powers of 2.
+				while (chip.Bank > _bankMask)
+					_bankMask = unchecked((byte) ((_bankMask << 1) | 1));
+				
+				var bank = new byte[BankSize];
+
+				bank.AsSpan().Fill(DummyData);
+				chip.ConvertDataToBytes().CopyTo(bank.AsSpan());
+
+				_banks[chip.Bank] = bank;
+
+				if (chip.Bank > maxBank)
+					maxBank = chip.Bank;
 			}
 
-			// for safety, initialize all banks to dummy
-			for (var i = 0; i < _banks.Length; i++)
-			{
-				_banks[i] = dummyBank;
-			}
-
-			// now load in the banks
-			for (var i = 0; i < count; i++)
-			{
-				if (newAddresses[i] == 0x8000)
-				{
-					_banks[newBanks[i] & _bankMask] = newData[i];
-				}
-			}
+			_bankCount = maxBank + 1;
 
+			// Start with bank 0.
 			BankSet(0);
 		}
 
+		public override IEnumerable<MemoryDomain> CreateMemoryDomains()
+		{
+			yield return new MemoryDomainDelegate(
+				name: "ROM",
+				size: _bankCount * BankSize,
+				endian: MemoryDomain.Endian.Little,
+				peek: a => _banks[a >> 13][a & 0x1FFF],
+				poke: (a, d) => _banks[a >> 13][a & 0x1FFF] = d,
+				wordSize: 1
+			);
+		}
+
 		protected override void SyncStateInternal(Serializer ser)
 		{
-			ser.Sync("BankMask", ref _bankMask);
 			ser.Sync("BankNumber", ref _bankNumber);
 			ser.Sync("ROMEnable", ref _romEnable);
 
 			if (ser.IsReader)
-			{

```

Comment:
idrc if you include unrelated changes here, but please use braces XOR single-line for conditionals.

Reply:
Can you elaborate what that means?

Reply:
```csharp
if (ser.IsReader)
{
	BankSet(_bankNumber | (_romEnable ? 0x00 : 0x80));
}
```
or
```csharp
if (ser.IsReader) BankSet(_bankNumber | (_romEnable ? 0x00 : 0x80));
```
but not
```csharp
if (ser.IsReader)
	BankSet(_bankNumber | (_romEnable ? 0x00 : 0x80));
```
or
```csharp
if (ser.IsReader) { BankSet(_bankNumber | (_romEnable ? 0x00 : 0x80)); }
```

Reply:
Thank you for the clarification- I will update the styling in a commit tonight.